### DISCUSSION TITLE
keyword continue for methods to make code more clean

### DISCUSSION THREAD

**Discussion by amrmostafa800** [dotnet/csharplang#7798] [Language Ideas]
edit : i understand this will been really bad design (edited)
what about make it like normal method and only when return in it like return in original method
sorry for bad english
	
```
   static async Task<int> OriginalMethodAsync()
    {
        bool isOk = true;
        await ContinueMethod(number);
        return 42;
    }

    static continue async Task<int> ContinueMethod(bool result) // its return type must been same Caller method return type
    {
		if(result)
		{
		 return 8; // if return here like return in Caller Method
		}
		 // if dont return Caller Method will continue normal
    }
```

**Comment by HaloFour** [comment]
See: #1413

**Reply by amrmostafa800** [reply]
i think not same idea bycouse it recive parameter but i want to can call local variable of original method in it - i edit to make more clear

**Reply by HaloFour** [reply]
It still falls under the umbrella of method inlining or macros.

**Reply by amrmostafa800** [reply]
can please give me example to do same thing in example up with inlining or anything in c# ?

**Reply by HaloFour** [reply]
You can't, C# doesn't have a feature like this.  Method inlining is a feature of the runtime and the team has not expressed interest in adding macros.  My opinion is that this feature would not be considered either, and that the ability to reference identifiers that do not actually exist within the method would not be considered a positive thing.

**Reply by amrmostafa800** [reply]
ok thank you cant do this using code generation ? and make compiler ignore continue method (bycouse code generator will convert it and original method to one method) ? to do this at build time (i still dont know how code generation work but i think maybe can do thing like this using it ?)

**Reply by HaloFour** [reply]
A source generator can emit additional methods into the current class, assuming it's marked as `partial`.

There is another proposal to support method interception, which would let you replace one method invocation with another: https://github.com/dotnet/csharplang/issues/7009

Neither of these options let you rewrite the source within a specific method, though.

**Reply by amrmostafa800** [reply]
ok thanks for help

**Comment by TahirAhmadov** [comment]
I think you want a complicated operation broken up into multiple methods, while sharing the state of the overall operation between these methods. I doubt this has much to do with inlining or performance per se.
The way I understand your request - it's a regular (not necessarily a very frequent) pattern. Currently I declare a type to hold the state, most likely a struct, then pass it around by ref between the methods in your group which form the single "big" operation. At run time, all the overhead is optimized away and the whole thing works just like if you wrote a giant method - while allowing you to write code in a more modular manner, which has tons of benefits. The downside here is that you lose (get degraded) the language assignment and NRT analysis. 
You also want to consider extracting parts of the functionality into completely separate types - this may not be always possible, but it's the first step. 

PS. There are also `#region` directives which can help some times. 
PPS. Also local functions may help. The advantage of the latter two approaches is that you get the full flow analysis; the downside is that you still have a huge method - it's somewhat more readable/ maintainable - but I don't like "huge" methods regardless of the "lipstick" you put on it.

**Reply by amrmostafa800** [reply]
Thanks, not because of performance, but I've realized that the code might be clearer this way, making it easier to read. Yes, using regions or even comments can help. I'm aware of 'ref,' currently, I don't have a method large enough to necessitate such a thing.

But When attempting to divide the code into parts for better clarity, you might need, for example, an if statement to verify the result of the method. I thought it might be a good idea to have something like that, though it may not be as effective as I anticipate. After all, I'm a new developer, and it might lead to issues like code duplication if implemented exactly as I explained.

However, in the end, I decided that his suggestion might be a good idea. If it were to be implemented, they would likely manage it well, not exactly as I explained.

**Reply by amrmostafa800** [reply]
ok i understand you now - most important part is return (when return to this method like return in original method)

**Reply by TahirAhmadov** [reply]
For `return` situations like that, you need a wrapper return type on the "child" methods like `Result<T>` - basically:
```cs
int Foo()
{
  var state = new FooState { ... };
  if(Foo_PartA(ref state) is {} t && t.HasResult) return t.Result;
  return 456;
}
struct FooState
{
}
Result<int> Foo_PartA(ref FooState state)
{
  return new(123);
}
```

**Reply by amrmostafa800** [reply]
thanks but this not same i mean (i only ask for it for more clean code) i edit up - you thing this good or still bad idea ?
and thanks for tell me about Result class

**Reply by TahirAhmadov** [reply]
I think given that my code snippet is already valid language, it's very unlikely that they would invest into making it even simpler in the direction of what you are proposing. Especially so because while it is a pattern that I have to deal with from time to time, it's **not a frequent** pattern.
Also, there are questions about your proposal, like how do you link these methods together in a "group"?

**Reply by amrmostafa800** [reply]
i mean method can only been inside other method direct cant been inside () can used for validation (i know can just make it inside if but it look good more with this why i think ?)

the resson is can used in asp.net controller for clean validation (without need to make if) validation method will do everything

**Reply by amrmostafa800** [reply]
maybe can do it when Interceptors released as package
